### ISSUE TITLE
importlib.reload not updated when only change a little bit code layout

### ISSUE THREAD

**Issue by BinBrian** [python/cpython#120932]
# Bug report

### Bug description:

```python
import os
import importlib

TEST_MODULE_NAME = "tmod"


def WriteCodeContext(text: str, filePath):
    with open(filePath, "w") as f:
        f.write(text)
        f.flush()


scriptDir = os.path.abspath(os.path.dirname(__file__))
testFilePath = os.path.join(scriptDir, f"{TEST_MODULE_NAME}.py")

source = """
a = 1
"""

target = """
a = 2
"""

WriteCodeContext(source, testFilePath)
mod = importlib.import_module(f"{TEST_MODULE_NAME}")

old = mod.a

WriteCodeContext(target, testFilePath)
importlib.reload(mod)

new = mod.a

assert old != new  # failed, 1 != 1
```
When I added a small number of comments, the results changed
```python

# ... ...

target = """
a = 2 # some comments
"""

# ... ...

assert old != new  # pass, 1 != 2
```

### CPython versions tested on:

3.8

### Operating systems tested on:

Windows

### Linked PRs
* gh-121620

**Comment by ZeroIntensity** [comment]
This is confirmed on Python 3.12 and the CPython main branch. Here's a smaller reproducer:
```py
import importlib
from pathlib import Path

MODULE_NAME = "tmod"
MODULE_PATH = Path(MODULE_NAME + ".py")


with open(MODULE_PATH, "w") as f:
    f.write(
        """
a = 1
"""
    )

mod = importlib.import_module(MODULE_NAME)
old = mod.a

with open(MODULE_PATH, "w") as f:
    f.write(
        """
a = 2
"""
    )

mod = importlib.reload(mod)
new = mod.a

assert old != new  # failed, 1 != 1
```

At first glance, this looks like a problem related to cache invalidation inside `importlib`, which would explain why it doesn't occur when comments are added. I'm guessing it does this by the size of the file, since this problem only occurs when the original file size is equivalent to the new one (which is rare, but still a bug).


I played around with this a little bit, and it seems likely that the problem is with `SourceLoader.get_code` -- only the file is updated, not the compiled `pyc` file. `get_code` then reads that old compiled file, and ignores the updated `py` file.

I'll submit a patch for this ASAP.

**Comment by ZeroIntensity** [comment]
The problem is [here](https://github.com/python/cpython/blob/6557af669899f18f8d123f8e1b6c3380d502c519/Lib/importlib/_bootstrap_external.py#L747). The repro updates the file too quickly for the number to be different, and when the size is the same, it doesn't get caught there either.

I'm not all too sure this is worth fixing, considering it will probably have negative performance benefits, and only happens in very extreme cases. It might be worth documenting, though. What do you think, @serhiy-storchaka (pinging you considering you helped on the previous `zipimport` issue)?

**Comment by serhiy-storchaka** [comment]
This is not related to `zipimport`, but it looks like the same issue as #121376 (there may be other open issues for the same problem).

**Comment by serhiy-storchaka** [comment]
Repeating myself from https://github.com/python/cpython/issues/121376#issuecomment-2222088675:

> it can be worked around by considering the .py file over bytecode when the timestamp second value is the same, at the cost of sometimes not using bytecode when it could have been

This is possible, but this may be not so simple. Currently the code compares the timestamp of the `.py` file (truncated to seconds) with the timestamp saved in the `.pyc` file. They are supposed to be equal. You need a third timestamp to handle the case when the `.py` file was changed within a second.

* Use the timestamp of the `.pyc` file. It needs to be strongly larger than the timestamp of the `.py` file (truncated to seconds) to ensure that the `.pyc` file is not outdated. But additional `stat()` call adds an overhead. We are trying to minimize the number of system calls for import.
* Use the current time. The problem is that on the FAT filesystem the file timestamp has granularity of 2 seconds. So the timestamp of just written file will often be different from the current time truncated to seconds.. And you cannot easily say what is the granularity of the file timestamp. There is also lesser problem of double rounding on filesystems with more precise file timestamps.
* Hybrid approach. If the cost of `stat()` is significant, we can use the current time, and only if the difference between timestamps is small, use the timestamp of the `.pyc` file. It will be not easy to test.

**Comment by serhiy-storchaka** [comment]
#121620 implements this idea.

**Comment by ZeroIntensity** [comment]
> This is not related to `zipimport`, but it looks like the same issue as #121376 (there may be other open issues for the same problem).

Right, I was just assuming you had knowledge of the import system :)
